README
README file for yepos

yepos is another dictionary program for PalmOS.
it's primary features are:
(0) RMS-sense freedom, you can use and distribute it under
    the terms of the GNU GPLv3+
(1) it is older devices-oriented:
    it runs on PalmOS 2.0 devices with 16 MHz DragonBall quite fluently
    (unfortunately, they can't bear any really large dictionary);
    generally, one of the most important goals is to to minimise
    the power consumption;
    it is written with badly adjusted digitizer in mind
(2) it's host-side programs are written in C within C89 standard
    (apart from dependency on zlib), single object file per program,
    so should be very portable and easy to build
(3) it has both database compiler and decompiler to regenerate
    the database source file bit-to-bit
(4) compressed databases support (via SysZLib)
(5) minimalistic user interface, e.g. find action is triggered with
    carriage return stroke or `Find' hardware button; articles are
    scrolled like in Weasel with tap on upper/lower half of the screen;
    no form titles, no fancy buttons, no other non-functional decorations

Some not so important features that are also planned:
(1) databases in memory cards as well as in Palm RAM
(2) reverse search (in articles bodies rather than in their titles)

Not planned neither in foreseeable future, nor later:
(0) multibyte encodings

There are other free dictionary programs available for PalmOS.
Here I list reasons why they didn't suit me
(0) PalmOpenDic doesn't come with database decompiler;
    it's host tools are written in Java
    and I am not familiar with this language;
    the most important - it is very intellectual, but it is slow.
(2) PAdict is a wonderful work oriented on Japaneese
(3) PocketBabylon can't read compressed databases

yepos is free program; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.

Although yepos is free, PalmOS is not, so it is not very ethical to
support Palm producers by writing free programs that primarily run on their
proprietary system. yepos author writes the program for personally
historical reasons (another excuse is that PalmOS is not supported by
it's developers any more).

Please don't use proprietary programs when you have a free alternative!

Another necessary copyright note: files in extern/ directory are neigher
copyrighted by yepos author, nor parts of yepos; they are used with it
as external libraries. yepos authors distrubute them under the same terms
as yepos. you also can get these files under their original
(more permissive) licenses from their original distributors.
for precise copyright statements see the respective files.


ChangeLog
This file is produced automatically from git repository
You may want to edit it manually if you distribute
your changes without accessing git repository

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Wed Jul 15 20:26:16 2009 +0400

    bump version to 0.2-rc1

* src/Makefile

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Mon Jun 1 14:10:18 2009 +0400

    approach to ancient hosts

    in particular, make utils buildable under Quasijarus
    (CFLAGS="-DNO_MALLOC_DECL -DNO_REALLOC_DECL -DNO_REMOVE" make -e)

* doc/en/yepos.html
* doc/ru/yepos.html
* doc/yepos.html
* util/Makefile
* util/deyepos.c
* util/dict2y.c
* util/dynamic_buffer.h
* util/yeposc.c
* util/ysort.c

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Sun May 31 19:52:44 2009 +0400

    some more documentation

* doc/en/yepos.html
* doc/ru/yepos.html
* doc/yepos.html

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Sun May 31 19:14:10 2009 +0400

    fix some corner cases for article search

* src/yepos.c

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Fri May 29 15:11:13 2009 +0400

    fix compare_item

    it should be asymmetric: if search word
    shorter than article title it should return 0

* src/yepos.c

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Thu May 28 20:56:07 2009 +0400

    switch off benchmarks

* src/settings.c
* src/yepos.c

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Thu May 28 20:55:22 2009 +0400

    bump version to post-0.2-rc0

* src/Makefile

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Tue May 26 21:10:52 2009 +0400

    regen ChangeLog

* ChangeLog

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Tue May 26 20:41:06 2009 +0000

    bump version to 0.2-rc0

* src/Makefile

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Tue May 26 19:20:41 2009 +0000

    update documentation

* README
* doc/en/yepos.html
* doc/ru/yepos.html
* doc/yepos.html

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Thu May 21 07:28:30 2009 +0400

    implement sort tables (yepos, yeposc, deyepos)

* include/signs.h
* src/dict_header.c
* src/dict_header.h
* src/yepos.c
* util/deyepos.c
* util/yeposc.c

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Tue May 19 07:15:29 2009 +0400

    fix spaces

* util/yeposc.c

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Mon May 18 08:06:19 2009 +0400

    mv dynamic_buffer to separate header

* util/Makefile
* util/dict2y.c
* util/dynamic_buffer.h
* util/ysort.c

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Thu May 14 16:52:50 2009 +0400

    add ysort

    sort yeposc input file according to sort table
    (embedded or default)

    Usage: ysort unsorted.txt sorted.txt

    Current implementation loads the whole file into RAM;
    this is probably not a problem
    since the largest file I tested (about 110 MB)
    can be easily processed by any newer machine

* util/Makefile
* util/data/russian.koi8
* util/ysort.c

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Sat May 9 19:05:14 2009 +0000

    add star2y

    convert stardict2txt output into yeposc input format via sed(1)
    Usage:
    stardict2txt serb2croat.ifo
    sed -f star2y.sed <serb2croat.txt >s2c_y.txt
     where
      stardict2txt is a program coming with stardict-tools distribution;
      serb2croat.ifo is stardict database description file name;
       (there also should be serb2croat.dict and serb2croat.idx files:
        typical uncompression command is
        `gzip -dc <serb2croat.dict.dz >serb2croat.dict')

* util/star2y.sed

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Sat May 9 16:56:10 2009 +0000

    add dict2y

    convert DICT databases to yeposc input format;
    uses dictunformat utility coming with dictd;
    typical usage is
     dictunformat serb2croat.index<serb2croat.dict | dict2y>serb2croat.txt

* util/Makefile
* util/dict2y.c

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Thu May 7 15:06:15 2009 +0400

    rm odd break

* src/yepos.c

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Fri May 8 22:03:06 2009 +0000

    fix comment

* util/deyepos.c

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Thu May 14 16:22:16 2009 +0400

    update TODOs

* README

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Mon May 18 07:38:32 2009 +0400

    typo fix

* README

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Thu May 7 15:05:36 2009 +0400

    replace email with ineiev@users.berlios.de

* cat_changelog

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Sat Mar 21 13:43:53 2009 +0300

    regenerate ChangeLog

* ChangeLog

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Sat Mar 21 13:41:55 2009 +0300

    rm blank-only lines

* cat_changelog

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Sat Mar 21 13:17:21 2009 +0300

    note on moral base

* doc/en/yepos.html
* doc/ru/yepos.html
* doc/yepos.html

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Sat Mar 21 13:16:22 2009 +0300

    typo

* doc/yepos.html

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Sat Mar 21 13:15:03 2009 +0300

    note on SysZLib installation

* doc/en/yepos.html
* doc/ru/yepos.html
* doc/yepos.html

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Tue Mar 17 13:59:48 2009 +0300

    update note on PalmOS slavery

* README

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Tue Mar 17 13:25:22 2009 +0300

    regenerate ChangeLog

* ChangeLog

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Tue Mar 17 13:23:47 2009 +0300

    initial add

* cat_changelog

Author: D.T.Ineiev <ineiev@users.berlios.de>
Date:   Fri Feb 6 16:49:36 2009 +0300

    migrate to BerliOS

* doc/en/yepos.html
* doc/ru/yepos.html
* doc/yepos.html
* include/signs.h
* src/Makefile
* src/control_ids.h
* src/dict_header.c
* src/dict_header.h
* src/enums.h
* src/globals.h
* src/mem_ory.c
* src/mem_ory.h
* src/mem_ory.test.c
* src/prefs.c
* src/prefs.h
* src/settings.c
* src/settings.h

src/settings.c
#include<PalmOS.h>/*yepos the PalmOS dictionary: settings form functions*/
#include"../include/signs.h"
#include"settings.h"
#include"control_ids.h"
#include"dict_header.h"
#include"enums.h"
#include"globals.h"
#include"prefs.h"
#include"show_battery.h"
enum local_constants{info_line_inc=4};
enum debug_constants{facunde=0};
static int upper_db,selected,change_flag,info_line_no,comment_top_reached;
static struct dict_header dh;
static ControlType*
get_control_ptr(FormType*f,int id)
{return FrmGetObjectPtr(f,FrmGetObjectIndex(f,id));}
static void
show_buttons(FormType*f)
{int n,m,i=0;struct database_handle**h;ControlType*b;
 h=get_database_list(&n);n-=upper_db;
 m=Dictionary_Pushbuttons_number;if(m>n)m=n;
 if(upper_db)
 {b=get_control_ptr(f,Dictionary_Pushbutton_First_id);
  CtlSetLabel(b,"...");CtlSetValue(b,0);i++;
 }
 for(;i<m;i++)
 {b=get_control_ptr(f,Dictionary_Pushbutton_First_id+i);
  CtlSetLabel(b,h[i+upper_db]->name);
  CtlSetValue(b,selected-upper_db==i);
 }
 if(m==n)for(;i<Dictionary_Pushbuttons_number;i++)
 {b=get_control_ptr(f,Dictionary_Pushbutton_First_id+i);
  CtlHideControl(b);
 }else
 {b=get_control_ptr(f,Dictionary_Pushbutton_Last_id);
  CtlSetLabel(b,"...");CtlSetValue(b,0);
 }
}static int
load_selected_header(void)
{int n;struct database_handle**h=get_database_list(&n);
 if(selected>n)return!0;if(dh.db)close_dict_header(&dh);
 return load_dict_header(&dh,h[selected]->card,h[selected]->id);
}static void
draw_tail(int visible,int y)
{if(visible)WinDrawChars("...",3,0,y);
 else WinDrawChars("      ",6,0,y);
}static int
draw_resource_string(int id,int x,int y)
{MemHandle mh=DmGetResource('tSTR',id);const char*s;
 if(!mh)return!0;s=MemHandleLock(mh);if(!s)return!0;
 WinDrawChars(s,StrLen(s),0,y);MemHandleUnlock(mh);return 0;
}static int
comment_area_top(void)
{int n;get_database_list(&n);if(n>5)n=5;return 12+n*14;}
static void
clr_db_comment(void)
{static struct RectangleType r={{0,0},{screen_width,0}};
 int y;y=comment_area_top()+1;
 r.topLeft.y=y;r.extent.y=137-y;WinEraseRectangle(&r,0);
}static void
show_selected_database(FormType*f)
{int i,y,comment_size,incy=11,max_y=130,line=0;
 const char*comment;y=comment_area_top();
 if(info_line_no<=line)
 {if(!draw_resource_string(Dictionary_No_Str_id,0,y))
  {char s[17];StrIToA(s,selected);WinDrawChars(s,StrLen(s),100,y);
   y+=incy;
  }
 }line++;
 if(info_line_no<=line)
 {if(!draw_resource_string(Db_Features_Str_id,0,y))
  {char s[17];StrIToH(s,dh.features);s[3]='x';
   WinDrawChars(s+2,6,100,y);y+=incy;
  }
 }line++;
 comment=dh.comment;comment_size=dh.comment_size;
 for(i=0;0<comment_size;i++)
 {if(comment[i]=='\n'||i==comment_size)
  {if(line++>=info_line_no)
   {WinDrawChars(comment,i,0,y);y+=incy;}
   comment+=i+1;comment_size-=i+1;i=-1;
  }if(y>max_y)break;
 }draw_tail(comment_size>0,135);comment_top_reached=comment_size<=0;
}static void
return_from_settings(void)
{if(!change_flag)skip_next_redraw();
 if(dh.db)close_dict_header(&dh);at_close_app(0);goto_form(MainForm_id);
}static void
on_enter(void)
{FormType*f;unsigned long tic=TimGetTicks();int n;
 f=get_current_form();selected=get_current_db_idx();get_database_list(&n);
 if(n>Dictionary_Pushbuttons_number&&selected>Dictionary_Pushbuttons_number-1)
 {upper_db=selected-Dictionary_Pushbuttons_number/2;
  if(upper_db+Dictionary_Pushbuttons_number>n)
   upper_db=n-Dictionary_Pushbuttons_number;
 }
 show_buttons(f);FrmDrawForm(f);show_battery(!0);
 at_close_app(return_from_settings);
 if(!load_selected_header())show_selected_database(f);
 if(facunde)
 {char s[17];StrIToA(s,TimGetTicks()-tic);
  WinDrawChars(s,StrLen(s),screen_width-StrLen(s)*5-2,-2);
 }change_flag=info_line_no=0;
}static int
process_push(int n)
{FormType*f=get_current_form();ControlType*b;
 int db_num,i;get_database_list(&db_num);
 n-=Dictionary_Pushbutton_First_id;
 if(upper_db+n==selected&&(n||!upper_db)
    &&(upper_db+Dictionary_Pushbuttons_number>db_num))
  return 0;
 if(upper_db&&!n)
 {b=get_control_ptr(f,Dictionary_Pushbutton_First_id);
  CtlSetValue(b,0);
  upper_db-=Dictionary_Pushbuttons_number-3;if(upper_db<0)upper_db=0;
  b=get_control_ptr(f,Dictionary_Pushbutton_Last_id);
  CtlSetValue(b,0);
  show_buttons(f);return!0;
 }
 if(db_num-upper_db>Dictionary_Pushbuttons_number
  &&n==Dictionary_Pushbuttons_number-1)
 {b=get_control_ptr(f,Dictionary_Pushbutton_Last_id);
  CtlSetValue(b,0);
  upper_db+=Dictionary_Pushbuttons_number-3;
  if(upper_db>db_num-Dictionary_Pushbuttons_number)
   upper_db=db_num-Dictionary_Pushbuttons_number;
  b=get_control_ptr(f,Dictionary_Pushbutton_First_id);
  CtlSetValue(b,0);
  show_buttons(f);return!0;
 }
 for(i=0;i<Dictionary_Pushbuttons_number;i++)
 {b=get_control_ptr(f,Dictionary_Pushbutton_First_id+i);
  CtlSetValue(b,i==n);
 }selected=upper_db+n;
 if(!load_selected_header())
 {info_line_no=0;clr_db_comment();show_selected_database(f);}
 return 0;
}static int
process_pendown(int x,int y)
{int y0=comment_area_top(),y1=135;if((y>y1&&x>20)||y<y0)return 0;
 if(y>(y1+y0)/2)
 {if(!comment_top_reached)
  {info_line_no+=info_line_inc;clr_db_comment();
   show_selected_database(get_current_form());
  }
 }else if(info_line_no)
 {info_line_no-=info_line_inc;if(info_line_no<0)info_line_no=0;
  clr_db_comment();show_selected_database(get_current_form());
 }return!0;
}Boolean
settings_form_handler(EventType*e)
{switch(e->eType)
 {case frmOpenEvent:on_enter();break;
  case penDownEvent:return process_pendown(e->screenX,e->screenY);
  case ctlSelectEvent:
   switch(e->data.ctlSelect.controlID)
   {case OK_Button_id:
     if(get_current_db_idx()!=selected)
     {load_database(selected);change_flag=!0;}
     return_from_settings();break;
    case Cancel_Button_id:
     return_from_settings();break;
    case About_Button_id:FrmHelp(About_Str_id);break;
    default:
     if(e->data.ctlSelect.controlID>=Dictionary_Pushbutton_First_id
        &&e->data.ctlSelect.controlID<=Dictionary_Pushbutton_Last_id)
      process_push(e->data.ctlSelect.controlID);
     else return 0;
   }break;
  default:return 0;
 }return!0;
}/*Copyright (C) 2008, 2009 Ineiev<ineiev@users.berlios.de>, super V 93

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.*/
